File: java/0039_combinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> comb = new ArrayList<Integer>();
        backtrack(res,comb,candidates,target,0);
        return res;
    }
    public void backtrack(List<List<Integer>> res, List<Integer> comb, int[] candidates,int target,int idx)
    {
        if(idx>=candidates.length)
            return;
        if(target==0)
        {
            res.add(new ArrayList<Integer>(comb));
            return;
        }
        if(target-candidates[idx]>=0)
        {
            comb.add(candidates[idx]);
            backtrack(res,comb,candidates,target-candidates[idx],idx);
            comb.remove(comb.size()-1);
        }
        else
            return;
        backtrack(res,comb,candidates,target,idx+1);
    }
}

File: cpp/0039_combinationSum.cpp
// backtrack(candidates,target,idx)代表当前已经搜索到candidates的第idx个位置
// 1.不搜索第idx，搜索下一个
// 2.搜索idx
// 对candidates排序，如果已经大于target，直接跳出
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<vector<int>> res;
        vector<int> comb;
        backtrack(res,comb,candidates,target,0);
        return res;
    }
    void backtrack(vector<vector<int>>&res, vector<int> &comb, vector<int> &candidates,int target,int idx)
    {
        if(idx>=candidates.size())
            return;
        if(target==0)
        {
            res.push_back(comb);
            return;
        }
        if(target-candidates[idx]>=0)
        {
            comb.push_back(candidates[idx]);
            backtrack(res,comb,candidates,target-candidates[idx],idx);
            comb.pop_back();
        }
        else
            return;
        backtrack(res,comb,candidates,target,idx+1);
    }
};
